swarm/src/lib.rs

```diff
@@ -340,6 +341,9 @@ where
     /// Multiaddresses that our listeners are listening on,
     listened_addrs: HashMap<ListenerId, SmallVec<[Multiaddr; 1]>>,
 
+    /// Cache of peers with their known external addresses.
+    peer_addrs: PeerAddresses,
```

Comment:
I don't think the `Swarm` itself should keep a cache of addresses, I'd rather leave this to the individual modules.